worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,874 @@
+/*
+	Copyright 2019 Supercomputing Systems AG
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+		http://www.apache.org/licenses/LICENSE-2.0
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+use super::{db::SidechainDB, Error, Result};
+use codec::{Decode, Encode};
+use log::*;
+use rocksdb::WriteBatch;
+use sp_core::H256;
+use std::{collections::HashMap, fmt::Debug, hash::Hash, path::PathBuf};
+use substratee_worker_primitives::{
+	block::{BlockHash, BlockNumber},
+	traits::{Block as BlockT, SignedBlock as SignedBlockT},
+};
+/// key value of sidechain db of last block
+const LAST_BLOCK_KEY: &[u8] = b"last_sidechainblock";
+/// key value of the stored shards vector
+const STORED_SHARDS_KEY: &[u8] = b"stored_shards";
+
+/// ShardIdentifier type
+type ShardIdentifier<A> = <<A as SignedBlockT>::Block as BlockT>::ShardIdentifier;
+/// Helper struct, contains the blocknumber
+/// and blockhash of the last sidechain block
+#[derive(PartialEq, Eq, Clone, Encode, Decode, Debug, Default)]
+pub struct LastSidechainBlock {
+	/// hash of the last sidechain block
+	pub hash: H256,
+	/// block number of the last sidechain block
+	pub number: BlockNumber,
+}
+
+/// Struct used to insert newly produced sidechainblocks
+/// into the database
+pub struct SidechainStorage<SignedBlock: SignedBlockT>
+where
+	SignedBlock: Encode + Decode,
+	<<SignedBlock as SignedBlockT>::Block as BlockT>::ShardIdentifier:
+		Hash + Eq + Debug + Encode + Decode + Copy,
```

Comment:
This should not be needed anymore after putting the trait bounds on the associated type.

Reply:
Removed